graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/GraylogPreflightGeneratePeriodical.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.bootstrap.preflight;
+
+import org.graylog.security.certutil.cert.storage.CertMongoStorage;
+import org.graylog.security.certutil.csr.CsrSigner;
+import org.graylog.security.certutil.csr.storage.CsrMongoStorage;
+import org.graylog.security.certutil.csr.storage.CsrStorage;
+import org.graylog2.cluster.NodePreflightConfig;
+import org.graylog2.cluster.NodePreflightConfigService;
+import org.graylog2.plugin.periodical.Periodical;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+
+@Singleton
+public class GraylogPreflightGeneratePeriodical extends Periodical {
+    private static final Logger LOG = LoggerFactory.getLogger(GraylogPreflightGeneratePeriodical.class);
+
+    private final NodePreflightConfigService nodePreflightConfigService;
+
+    private final CsrStorage csrStorage;
+    private final CertMongoStorage certMongoStorage;
+
+    private String caKeystoreFilename = "datanode-ca.p12";
+    private PrivateKey caPrivateKey;
```

Comment:
I'd consider keeping PrivateKey as field of a class a risky thing, especially when it is singleton.

Reply:
duly noted